### ISSUE TITLE
Include path handling is fragile.

### ISSUE THREAD

**Issue by olupton** [BlueBrain/CoreNeuron#715]
**Describe the issue**
Some header files, such as `nrnreport.hpp`, are copied to the build directory and appear both at:
```
{source}/coreneuron/io/reports/nrnreport.hpp
```
and
```
{build}/include/coreneuron/io/reports/nrnreport.hpp
```
and both of these can end up being found by `#include` directives during compilation.
This caused failures in the first draft of https://github.com/BlueBrain/CoreNeuron/pull/712 and is generally a bit fragile.

**Expected behaviour**
There should be a clear and enforced policy (e.g. `#pragma once` is banned), or the build system should be modified so that this ambiguity cannot arise. Perhaps these should never be in the build directory, but rather just the source and installation directories?

**Logs**
https://github.com/BlueBrain/CoreNeuron/runs/4490079420?check_suite_focus=true

**Comment by alkino** [comment]
It's not some, it's all: https://github.com/BlueBrain/CoreNeuron/blob/master/coreneuron/CMakeLists.txt#L331-L338

but why?

**Comment by olupton** [comment]
I assume it's basically because of the `nrnivmodl` workflow supporting running from both build and install directories, but I'm not sure.

Note that we also do things like copying source files to the build directory and compiling them from there (`partial_piv_lu.cu`).

**Comment by nrnhines** [comment]
I think that some of the features of nrnivmodl and its corresponding makefile can be adopted here. The former creates a ```makemod2c_inc``` file that the makefile includes which handles the location details for translating and compiling. e.g.
```
nrnivmodl modx
```
generates a ```x86_64/makemod2d_inc``` with
```
./invlfiresha.c: ../modx/invlfiresha.mod
        @printf " -> $(C_GREEN)NMODL$(C_RESET) $<\n"
        (cd "../modx"; MODLUNIT=$(NRNUNITS) $(NOCMODL) invlfiresha.mod -o "/home/hines/models/tqperf/x86_64")

./invlfiresha.o: ./invlfiresha.c
        @printf " -> $(C_GREEN)Compiling$(C_RESET) $<\n"
        $(COMPILE) -I"../modx" $(INCLUDES) -fPIC -c $< -o $@
```
in particular nothing is copied from the mod file folder, the translator is launched with the mod file folder as the working directory but the translated c file is created in the x86_64 folder,  and compiling makes sure any needed include files in the mod file folder are findable.

**Comment by pramodk** [comment]
@olupton : #712 has addressed this now?

**Comment by alkino** [comment]
Fixed by #712
